[aznude] Add new extractor

DirkGiggler opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds a new extractor for aznude.com. The site hosts sex scenes from movies. Since the clips are quite short, I believe they fall under Fair Use. Plus, they comply with DMCA requests ([see here](http://www.aznude.com/dmca.html)).

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,117 @@
+            'url': self._search_regex(r'(?:<a href=")(?P<url>(https?:)?//[^/]*\.aznude\.com/.*\.mp4)(?:"><div class="videoButtons"><i class="fa"></i>Download</div></a>)',
+                                      webpage,
+                                      'url',
+                                      fatal=True)
```
All formats should be extracted.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,117 @@
+                                  url,
+                                  default=None)
+
+        if (artist is not None) and (work is not None):
```
`_search_html_regex` instead.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,117 @@
+
+        return {
+            'id': video_id,
+            'title': title + " - " + numeric_id,
```
Point of using `numeric_id`?

DirkGiggler replied:
The title is common for all videos in the same movie/show for a particular artist. Due to a lack of numeric id of videos, I had to resort to using the last component of the url (e.g. `theblockbuster-smithdoe-hd-01-hd`). While that component is unique, it also has a lot of redundant information, resulting in filenames such as: 
`John Doe in The BlockBuster-theblockbuster-smithdoe-hd-01-hd.mp4` and 
`Jane Smith in Very Long Series-verylongseries-s13e12-smith-02.mp4`. So I created a numeric id from the last component, extracting video index as well as season+episode info, and appended it to the title, making the title unique as well as easy to understand. e.g. `John Doe in The BlockBuster - 01.mp4` and `Jane Smith in Very Long Series - s13e12-02.mp4`.

dstftw replied:
Add a test for such case.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,117 @@
+        url_prefix = parse_result.scheme + "://" + parse_result.netloc
+
+        entries = self._extract_entries(webpage, self._get_entry_regex(page_id), url_prefix)
+        return self.playlist_result(entries, page_id, self._get_webpage_title(webpage))
```
What's the point of this complication? Each clip has `show-clip` value for class attribute for `a`.
Replace with single playlist extractor based on this approach.

DirkGiggler replied:
Sorry, totally missed that. Will update along with the rest of the suggestions.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,117 @@
+
+
+class AZNudeIE(InfoExtractor):
+    IE_NAME = "aznude"
```
Single quotes.

DirkGiggler commented:
@dstftw I've updated the extractor as per your suggestions. Please take a look now.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,119 @@
+
+class AZNudeCollectionIE(InfoExtractor):
+    IE_NAME = 'aznude:collection'
+    _VALID_URL = r'https?://(?:www\.)?aznude\.com/(?P<id>(?:view|browse|tags)/.+\.html)'
```
It's not an id.

DirkGiggler replied:
Sorry, I took a shortcut there. I'll come up with a better regex.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,119 @@
+
+        entries = []
+        for path in re.findall(r'(?:<a[^>]+href=")(?P<url>[^"]+)(?:"[^>]+class="(?:[^"]+ )?show-clip(?:"| [^"]+")[^>]*>)', webpage):
+            entries.append( self.url_result(url_prefix + path, AZNudeIE.ie_key()) )
```
`utils.urljoin`.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,119 @@
+                format['format_id'] = 'HD'
+                format['quality'] = 2
+                format['resolution'] = '1280x720'
+                format['format_note'] = '720p video with 128 kbps audio'
```
Code duplication.

dstftw replied:
Audio is not 128.

DirkGiggler replied:
Apparently, the audio bitrate is not consistent for any quality level, probably because they're using VBR. I'll remove the audio bitrate from the description entirely.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,119 @@
+                format['format'] = 'High Quality'
+                format['format_id'] = 'HQ'
+                format['quality'] = 2
+                format['resolution'] = '640x360'
```
`width` and `height` instead.

dstftw commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,119 @@
+            'id': 'loiteringwithintent-mcnallytomei-hd-01-hd',
+            'ext': 'mp4',
+            'title': 'Marisa Tomei in Loitering With Intent - 01',
+            'thumbnail': 'https://cdn1.aznude.com/marisatomei/loiteringwithintent/LoiteringWithIntent-McNallyTomei-HD-01-gigantic-4.jpg',
```
No exact matches for URLs.

DirkGiggler replied:
I'm sorry, I don't understand.

dstftw replied:
Do not match exact URL.

DirkGiggler commented:
@dstftw Do we need to make further changes to the extractor?

dstftw commented:
Read my comments on what you should fix.

DirkGiggler commented:
@dstftw I have fixed them all. Kindly take another look.

dstftw commented:
You did not, at least not in this branch: code duplication not fixed, url_prefix not removed in favor of `urlojoin`.

timofonic commented:
@DirkGiggler  @dstftw  Any news?

DirkGiggler commented:
@dstftw Hey, sorry for abandoning the PR. I have updated the extractor and rebased my branch. I have also used this extractor for these past years now without any issues. The main addition I have had to make is that of checking for the `lightbox` attribute in the anchor tag. It seems to be a good way to distinguish the actual content from the (newly added) recommended content. Anyway, I request you to review the latest code. Thanks! :)

Lmn01234 commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,127 @@
+                format['quality'] = 1
+                format['width'] = 640
+                format['height'] = 360
+                format['format_note'] = '360p video with mono audio'
```
- - [ ]

Lmn01234 commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,127 @@
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        numeric_id = '-'.join(re.findall(r'(?P<num>(?:s\d+e\d+)|(?:\d+[xX]\d+)|(?:\d+))', video_id))
+        webpage = self._download_webpage(url, video_id)
```
[]()

Lmn01234 replied:
```suggestion
        webpage = self._download_webpage(url, video_id)
```

Lmn01234 commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,127 @@
+                format['width'] = 1280
+                format['height'] = 720
+                format['format_note'] = '720p video with stereo audio'
+            else:
```
**##**

Lmn01234 commented on youtube_dl/extractor/aznude.py:
```diff
@@ -0,0 +1,127 @@
+                                    webpage,
+                                    url,
+                                    default=None)
+        work = self._html_search_regex(r'in (?P<work><a href="/view/movie/[^/?]/[^/?]+\.html">[^<]+</a>)',
```
64
